File: test/unit/app_state_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:jordanian_ser_app/models/speaker.dart';
import 'package:jordanian_ser_app/services/app_state.dart';

void main() {
  group('AppState Tests', () {
    test('Initial state should have no gender and score 0', () {
      final state = AppState();
      expect(state.hasSelectedGender, isFalse);
      expect(state.gender, isNull);
      expect(state.score, equals(0));
    });

    test('Selecting male gender produces speakerId starting with M', () {
      final state = AppState();
      state.selectGender(Gender.male);
      expect(state.hasSelectedGender, isTrue);
      expect(state.gender, equals(Gender.male));
      expect(state.speakerId.startsWith('M'), isTrue);
    });

    test('Selecting female gender produces speakerId starting with F', () {
      final state = AppState();
      state.selectGender(Gender.female);
      expect(state.hasSelectedGender, isTrue);
      expect(state.gender, equals(Gender.female));
      expect(state.speakerId.startsWith('F'), isTrue);
    });

    test('Increment score increments correctly', () {
      final state = AppState();
      state.incrementScore();
      state.incrementScore();
      expect(state.score, equals(2));
    });

    test('Reset clears gender and resets score to 0', () {
      final state = AppState();
      state.selectGender(Gender.male);
      state.incrementScore();
      state.reset();
      expect(state.gender, isNull);
      expect(state.score, equals(0));
      expect(state.hasSelectedGender, isFalse);
    });
  });
}

File: test/unit/app_state_test.js
import { renderHook, act } from '@testing-library/react';
import { Gender } from '../models/speaker';
import { useAppState, AppStateProvider } from '../services/app_state';

describe('AppState Tests', () => {
  const wrapper = ({ children }) => <AppStateProvider>{children}</AppStateProvider>;

  test('Initial state should have no gender and score 0', () => {
    const { result } = renderHook(() => useAppState(), { wrapper });

    expect(result.current.hasSelectedGender).toBe(false);
    expect(result.current.gender).toBeNull();
    expect(result.current.score).toBe(0);
  });

  test('Selecting male gender produces speakerId starting with M', () => {
    const { result } = renderHook(() => useAppState(), { wrapper });

    act(() => {
      result.current.selectGender(Gender.male);
    });

    expect(result.current.hasSelectedGender).toBe(true);
    expect(result.current.gender).toBe(Gender.male);
    expect(result.current.speakerId.startsWith('M')).toBe(true);
  });

  test('Selecting female gender produces speakerId starting with F', () => {
    const { result } = renderHook(() => useAppState(), { wrapper });

    act(() => {
      result.current.selectGender(Gender.female);
    });

    expect(result.current.hasSelectedGender).toBe(true);
    expect(result.current.gender).toBe(Gender.female);
    expect(result.current.speakerId.startsWith('F')).toBe(true);
  });

  test('Increment score increments correctly', () => {
    const { result } = renderHook(() => useAppState(), { wrapper });

    act(() => {
      result.current.incrementScore();
      result.current.incrementScore();
    });

    expect(result.current.score).toBe(2);
  });

  test('Reset clears gender and resets score to 0', () => {
    const { result } = renderHook(() => useAppState(), { wrapper });

    act(() => {
      result.current.selectGender(Gender.male);
      result.current.incrementScore();
    });

    act(() => {
      result.current.reset();
    });

    expect(result.current.gender).toBeNull();
    expect(result.current.score).toBe(0);
    expect(result.current.hasSelectedGender).toBe(false);
  });
});